### ISSUE TITLE
Comparison between set and frozenset always evaluates to False

### ISSUE THREAD

**Issue by injust** [microsoft/pyright#10371]
**Describe the bug**

Pyright thinks there is no overlap between `set` and `frozenset`.

```
Condition will always evaluate to False since the types "set[str]" and "frozenset[str]" have no overlap  (reportUnnecessaryComparison)
```

**Code or Screenshots**

Code sample in [pyright playground](https://pyright-play.net/?strict=true&code=IYAgvCDeBEBmD29oF8BQAjcJYCd4C8BTAOwGdCAXACmAEpUAHHAS2OtDAnVqA)

```python
a = {"foo"}
b = frozenset(a)
print(a == b)
```

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

Detecting unnecessary comparisons is difficult in Python because the `__eq__` operator can be overridden. Pyright uses a heuristic to implement the `reportUnnecessaryComparison` check — it looks for types that overlap. For example, if you attempt to compare a `str` value to an `int` value, this will trigger this check.

There are cases where two classes have no apparent overlap but still implement an equality check. The `set` and `frozenset` are an example of this.

The `reportUnnecessaryComparison` check is not enabled by default. It's a strict-mode check partly because it has the potential for false positives. It may require you to make changes to your code to avoid these false positives.

You have a few options:
1. Leave the `reportUnnecessaryComparison` check disabled if you find that it's not providing enough value for you.
2. Use a `# type: ignore` or `# pyright: ignore` comment to suppress this particular check.
3. Convert one of the values to a type that has overlap (e.g. `a == set(b)`).

**Comment by jorenham** [comment]
> Detecting unnecessary comparisons is difficult in Python because the `__eq__` operator can be overridden. Pyright uses a heuristic to implement the `reportUnnecessaryComparison` check — it looks for types that overlap. For example, if you attempt to compare a `str` value to an `int` value, this will trigger this check.

There is no overlap between `frozenset` and `set`:

```pycon
>>> class ThawedSet(frozenset, set): ...
... 
Traceback (most recent call last):
  File "<python-input-2>", line 1, in <module>
    class ThawedSet(frozenset, set): ...
TypeError: multiple bases have instance lay-out conflict
```

So yes, it's indeed the case that `set[T] & frozenset[T]` is `Never`.

---

But that doesn't necessarily mean that `set[T].__eq__(frozenset[T])` is `Literal[False]`, which this counter-example shows:

```pycon
>>> {"a"} == frozenset({"a"})
True
```

The only conclusion I can logically draw from this, is that the reported `reportUnnecessaryComparison` error is a false positive in this case. 

So with that, should I read the "as designed" as a "won't fix", or are you willing to reconsider?

Also, I understand that it's an optional check we're talking about here. But personally, when I enable an optional check (that is*n't* marked as "experimental" or something), I expect it to accurately reflect the runtime behavior. I at least can't think of any other reason for why it would otherwise exist in the first place.

**Comment by erictraut** [comment]
I'm not interested in adding hard-coded special cases for this check. If you have a suggestion for how to improve a general check, let me know.

If you find the current check is too noisy, you can disable it.

**Comment by jorenham** [comment]
> I'm not interested in adding hard-coded special cases for this check. 

Fair enough.

> If you have a suggestion for how to improve a general check, let me know.

In typeshed, both `set` ([src](https://github.com/python/typeshed/blob/d270bb0dc1df3fcfae2808c9a5a0f7ec802e1fb2/stdlib/builtins.pyi#L1259)) and `frozenset` ([src](https://github.com/python/typeshed/blob/d270bb0dc1df3fcfae2808c9a5a0f7ec802e1fb2/stdlib/builtins.pyi#L1287)) override the `__eq__` method of `object`. Since it seems a bit redundant, there must be a good reason behind it. And as it turns out, these `__eq__` methods were added (in https://github.com/python/typeshed/pull/10464) to solve this very problem:

> I also added many missing `__eq__` methods. As well as their relevance to hashability, these are also used as a heuristic for mypy's overlapping-equality check.
